src/d3a/models/strategy/external_strategies/pv.py

```diff
@@ -0,0 +1,146 @@
+                list_offers_response_channel,
+                {"status": "error",
+                 "error_message": f"Error when listing offers on area {self.device_name}."})
+
+    def _offer(self, payload):
+        offer_response_channel = f'{self.device_name}/offer/response'
+        if not check_for_connected_and_reply(self.redis, offer_response_channel,
+                                             self.connected):
+            return
+        try:
+            arguments = json.loads(payload["data"])
+            assert set(arguments.keys()) == {'price', 'energy'}
+            arguments['seller'] = self.device_name
+            arguments['seller_origin'] = self.device_name
+        except Exception as e:
+            self.redis.publish_json(
+                offer_response_channel,
+                {"error": "Incorrect offer request. Available parameters: (price, energy)."}
+            )
+        else:
+            try:
+
+                offer = self.market.offer(**arguments)
+                self.redis.publish_json(offer_response_channel,
+                                        {"status": "ready", "offer": offer.to_JSON_string()})
+            except Exception as e:
+                logging.error(f"Error when handling offer create on area {self.device_name}: "
+                              f"Exception: {str(e)}, Offer Arguments: {arguments}")
+                self.redis.publish_json(
+                    offer_response_channel,
+                    {"status": "error",
+                     "error_message": f"Error when handling offer create "
+                                      f"on area {self.device_name} with arguments {arguments}."})
+
+    def _area_stats(self, payload):
+        area_stats_response_channel = f'{self.device_name}/stats/response'
+        if not check_for_connected_and_reply(self.redis, area_stats_response_channel,
+                                             self.connected):
+            return
+        try:
+            device_stats = {k: v for k, v in self.owner.stats.aggregated_stats.items()
+                            if v is not None}
+            market_stats = {k: v for k, v in self.area.stats.aggregated_stats.items()
+                            if v is not None}
+            self.redis.publish_json(
+                area_stats_response_channel,
+                {"status": "ready",
+                 "device_stats": device_stats,
+                 "market_stats": market_stats})
+        except Exception as e:
+            logging.error(f"Error reporting stats for area {self.device_name}: "
+                          f"Exception: {str(e)}")
+            self.redis.publish_json(
+                area_stats_response_channel,
+                {"status": "error",
+                 "error_message": f"Error reporting stats for area {self.device_name}."})
+
+    @property
+    def market(self):
+        return self.area.next_market
+
+    @property
+    def device_name(self):
+        return self.owner.name
```

Comment:
I think these two method could be moved to some Class ExternalStrategyBase as they are reimplemented in `PV` as well as in `Load`

Reply:
I suggest to postpone this for the following PR for the same story, because there is a lot of code duplication that can be avoided in these 2 classes. The reason why it is not that easy to do though, is because these classes do not inherit from ExternalStrategyBase, they inherit only from the LoadHours and PVStrategy.